### PR TITLE
luaPackages.lua-pam: init at unstable-2015-07-03

### PR DESCRIPTION
**Pull request by traxys**
## Description of changes

This allows to use PAM authentication in lua.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 277393` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>lua51Packages.lua-pam</li>
    <li>lua52Packages.lua-pam</li>
    <li>lua53Packages.lua-pam</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 277393` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by traxys** [comment]
Result of `nixpkgs-review pr 277393` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>lua51Packages.lua-pam</li>
    <li>lua52Packages.lua-pam</li>
    <li>lua53Packages.lua-pam</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 277393` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 277393` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>lua51Packages.lua-pam</li>
    <li>lua52Packages.lua-pam</li>
    <li>lua53Packages.lua-pam</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 277393` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>lua51Packages.lua-pam</li>
    <li>lua52Packages.lua-pam</li>
    <li>lua53Packages.lua-pam</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 277393` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>lua51Packages.lua-pam</li>
    <li>lua52Packages.lua-pam</li>
    <li>lua53Packages.lua-pam</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 277393` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>lua51Packages.lua-pam</li>
    <li>lua52Packages.lua-pam</li>
    <li>lua53Packages.lua-pam</li>
  </ul>
</details>

File: pkgs/top-level/lua-packages.nix [hunk 1]
```diff
@@ -59,6 +59,47 @@ rec {
...
+      hash = "sha256-YlMZ5mM9Ij/9yRmgA0X1ahYVZMUx8Igj5OBvAMskqTg=";
+      fetchSubmodules = true;
+    };
+
+    # The makefile tries to link to `-llua<luaversion>`
+    LUA_LIBS = "-llua";
+
+    buildInputs =
+      (lib.lists.optional stdenv.isLinux linux-pam)
+      ++ (lib.lists.optional stdenv.isDarwin openpam);
```
Comment (SuperSandro2000) [inline pkgs/top-level/lua-packages.nix:81]:
```suggestion
    buildInputs = lib.optional stdenv.isLinux linux-pam
      ++ lib.optional stdenv.isDarwin openpam;
```
lib.lists is an implementaion detail we shouldn't use. Also the () here are superfluous.

File: pkgs/top-level/lua-packages.nix [hunk 2]
```diff
@@ -59,6 +59,47 @@ rec {
...
+
+    installPhase = ''
+      runHook preInstall
+
+      install -d $out/lib/lua/${lua.luaversion}
+      install -m 755 pam.so $out/lib/lua/${lua.luaversion}/pam.so
```
Comment (SuperSandro2000) [inline pkgs/top-level/lua-packages.nix:87]:
```suggestion
      install -Dm755 pam.so $out/lib/lua/${lua.luaversion}/pam.so
```

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 277393` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>lua51Packages.lua-pam</li>
    <li>luaPackages.lua-pam (lua52Packages.lua-pam)</li>
    <li>lua53Packages.lua-pam</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 277393` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 277393` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

File: pkgs/top-level/lua-packages.nix [hunk 3]
```diff
@@ -59,6 +59,45 @@ rec {
   # a fork of luarocks used to generate nix lua derivations from rockspecs
   luarocks-nix = toLuaModule (callPackage ../development/tools/misc/luarocks/luarocks-nix.nix { });
 
+ lua-pam = callPackage({fetchFromGitHub, linux-pam, openpam}: buildLuaPackage rec {
+    pname = "lua-pam";
+    version = "unstable-2015-07-03";
+    # Needed for `disabled`, overridden in buildLuaPackage
+    name = "${pname}-${version}";
+
+    src = pkgs.fetchFromGitHub {
```
Comment (SomeoneSerge) [inline pkgs/top-level/lua-packages.nix:68]: fetchFromGiHub from callPackage over pkgs

File: pkgs/top-level/lua-packages.nix [hunk 4]
```diff
@@ -59,6 +59,45 @@ rec {
...
+      repo = "lua-pam";
+      rev = "3818ee6346a976669d74a5cbc2a83ad2585c5953";
+      hash = "sha256-YlMZ5mM9Ij/9yRmgA0X1ahYVZMUx8Igj5OBvAMskqTg=";
+      fetchSubmodules = true;
+    };
+
+    # The makefile tries to link to `-llua<luaversion>`
+    LUA_LIBS = "-llua";
+
+    buildInputs = lib.optional stdenv.isLinux linux-pam
```
Comment (SomeoneSerge) [inline pkgs/top-level/lua-packages.nix:79]: not a policy, but I optionals is arguably less error prone than optional

File: pkgs/top-level/lua-packages.nix [hunk 5]
```diff
@@ -59,6 +59,45 @@ rec {
   # a fork of luarocks used to generate nix lua derivations from rockspecs
   luarocks-nix = toLuaModule (callPackage ../development/tools/misc/luarocks/luarocks-nix.nix { });
 
+ lua-pam = callPackage({fetchFromGitHub, linux-pam, openpam}: buildLuaPackage rec {
+    pname = "lua-pam";
+    version = "unstable-2015-07-03";
+    # Needed for `disabled`, overridden in buildLuaPackage
+    name = "${pname}-${version}";
```
Comment (SomeoneSerge) [inline pkgs/top-level/lua-packages.nix:66]: just making sure, the correct handling of 'disabled' depends you manually defining the 'name'?

Comment (traxys) [reply pkgs/top-level/lua-packages.nix:66]: Yes :/

**Comment by SomeoneSerge** [comment]
@GaetanLepage you requested a review, bur I.m not familiar with either of lua or pam. I'll happily merge if someone involved with these subsystems approves

**Comment by GaetanLepage** [comment]
> @GaetanLepage you requested a review, bur I.m not familiar with either of lua or pam. I'll happily merge if someone involved with these subsystems approves

All right, sorry for the ping then.
 Thank you for your comments and suggestions anyway :)

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 277393` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>lua51Packages.lua-pam</li>
    <li>luaPackages.lua-pam (lua52Packages.lua-pam)</li>
    <li>lua53Packages.lua-pam</li>
  </ul>
</details>

**Review by SomeoneSerge — APPROVED** [review]
CC @NixOS/lua

Objections to merging tomorrow?

File: pkgs/top-level/lua-packages.nix [hunk 6]
```diff
@@ -59,6 +59,45 @@ rec {
...
+    installPhase = ''
+      runHook preInstall
+
+      install -Dm755 pam.so $out/lib/lua/${lua.luaversion}/pam.so
+
+      runHook postInstall
+    '';
+
+    # The package does not build with lua 5.4 or luaJIT
+    disabled = luaAtLeast "5.4" || isLuaJIT;
```
Comment (teto) [inline pkgs/top-level/lua-packages.nix:91]: disabled is actually meta.broken. You might as well assign it there.

**Review by teto — APPROVED** [review]

**Pull request merged**